Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp

```diff
@@ -0,0 +1,485 @@
+                extendedArgs.emplace_back("frameTime", AZ::Metrics::EventValue{ AZStd::in_place_type<AZ::s64>, 20 });
+                AZ::Metrics::CounterArgs counterArgs;
+                counterArgs.m_name = "Counter Event";
+                counterArgs.m_cat = "Test";
+                counterArgs.m_args = extendedArgs;
+                counterArgs.m_id = idString;
+                resultOutcome = googleTraceLogger.RecordCounterEvent(counterArgs);
+            }
+
+            {
+                // Record Async Start and End Events
+                // Also records the Async Instant event
+                constexpr AZStd::string_view asyncOuterEventName = "Async Event";
+
+                AZ::Metrics::AsyncArgs asyncArgs;
+                asyncArgs.m_name = asyncOuterEventName;
+                asyncArgs.m_cat = "Test";
+                asyncArgs.m_args = argsContainer;
+                asyncArgs.m_id = idString;
+                asyncArgs.m_scope = "Distinguishing Scope";
+                resultOutcome = googleTraceLogger.RecordAsyncEventStart(asyncArgs);
+
+                // Change the name of the event
+                asyncArgs.m_name = "Async Instant Event";
+                resultOutcome = googleTraceLogger.RecordAsyncEventInstant(asyncArgs);
+
+                // "Async Event" is the logical name of the being and end event being recorded
+                // So make sure the end event matches
+                asyncArgs.m_name = asyncOuterEventName;
+                resultOutcome = googleTraceLogger.RecordAsyncEventEnd(asyncArgs);
+            }
+        };
+
+        // Flush and closes the event stream
+        // This completes the json array
+        googleTraceLogger.ResetStream(nullptr);
+    }
+
+    BENCHMARK_F(JsonTraceEventLoggerBenchmarkFixture, BM_JsonTraceEventLogger_RecordEventsToInMemoryStream)(benchmark::State& state)
+    {
+        constexpr size_t eventOutputReserveSize = 4096;
+        AZStd::string metricsPayload;
+        // Reserve memory to prevent allocation of string data skewing results
+        // for byte container stream
+        metricsPayload.reserve(eventOutputReserveSize);
+        for ([[maybe_unused]] auto _ : state)
+        {
+            RecordMetricsToStream(AZStd::make_unique<AZ::IO::ByteContainerStream<AZStd::string>>(&metricsPayload));
+            metricsPayload.clear();
+        }
+    }
+
+    BENCHMARK_F(JsonTraceEventLoggerBenchmarkFixture, BM_JsonTraceEventLogger_RecordEventsToTemporaryDirFileStream)(benchmark::State& state)
+    {
+        // Open the file and truncate the file
+        constexpr AZ::IO::OpenMode openMode = AZ::IO::OpenMode::ModeWrite;
+        const AZ::IO::FixedMaxPath metricsFilepath = m_tempDirectory.GetDirectoryAsFixedMaxPath() / "metrics.json";
+
+        for ([[maybe_unused]] auto _ : state)
+        {
+            RecordMetricsToStream(AZStd::make_unique<AZ::IO::SystemFileStream>(metricsFilepath.c_str(), openMode));
+        }
+    }
```

Comment:
Writing metrics to a SystemFileStream is a lot slower than writing to a ByteContainerStream and then performing a single write to a SystemFileStream.

I am thinking of adding a buffered SystemFileStream class that maintains an internal buffer for the user

Reply:
Well a lot slower was an overstatement.
Turns out the opening of a file is accounts for most of the slowness of using a SystemFileStream.
The actual writing to the SystemFile is about 55% as fast as writing to an in-memory stream.

I also created a buffered SystemFile stream implementation locally and it turns out it is slower than just directly using SystemFileStream